### PR TITLE
Add a no dynamic linking build option

### PR DESCRIPTION
**Pull request by hoodmane**
To test dead code elimination.

### CODE REVIEW

**Comment by ryanking13** [comment]
Thanks @hoodmane! I'll review this soon.

It "dead code elimination" is the only reason you are trying to build no-dynamic linking version of Pyodide? Personally I am interested in no-dynamic linking version of Pyodide, which will make running packages like `pygame` easier.

**Comment by hoodmane** [comment]
> Is "dead code elimination" is the only reason you are trying to build no-dynamic linking version of Pyodide

Well yes the dramatically reduced load time is my main reason.

> pygame

One thing I've been thinking about is imitating componentize-py and statically linking together a bunch of so files.

**Comment by rth** [comment]
Interesting! What's the size of `pyodide.asm.wasm` with dynamic linking disabled? In one of the above commits, I see `Original size: 5781 KB` (instead of 8779 KB) which is not 10x smaller. I would have expected the load time to be somewhat linearly linked to the size of WASM. Unless dynamic linking adds some additional ovehead.

**Comment by hoodmane** [comment]
Now I'm seeing a much smaller difference. I also think we might be spending ~500ms or more of the startup time importing asyncio.

**Comment by rth** [comment]
Yes, asyncio is certainly rather large (around 0.5MB uncompressed) and so slow to parse and compile.

File: requirements.txt [hunk 1]
```diff
@@ -12,4 +12,4 @@
   pytest-cov
   pytest-httpserver
   pytest-benchmark
-  pytest-pyodide==0.53.0
+  git+https://github.com/hoodmane/pytest-pyodide@no-ctypes
```
Comment (ryanking13) [inline requirements.txt:15]: Could you release a new version of pytest-pyodide?

File: src/core/pre.js [hunk 2]
```diff
@@ -59,3 +59,5 @@ const errNoRet = () => {
     "Assertion error: control reached end of function without return",
   );
 };
+
+Module.reportUndefinedSymbols = () => {};
```
Comment (ryanking13) [inline src/core/pre.js:63]: Why is this required?

File: src/js/api.ts [hunk 3]
```diff
@@ -16,6 +16,9 @@ import { TypedArray } from "./types";
 // Exported for micropip
 API.loadBinaryFile = loadBinaryFile;
 
+// Exported for micropip
+API.loadBinaryFile = loadBinaryFile;
```
Comment (ryanking13) [inline src/js/api.ts:20]: Looks like there was a bad merge?

**Review by ryanking13 — COMMENTED** [review]
Thanks @hoodmane! I have some nitpicky comments first, and I will review this again when the CI gets green.

File: src/core/pre.js (hunk 2 above)
Comment (hoodmane) [reply src/core/pre.js:63]: In a nodylink build, `Module.reportUndefinedSymbols` is not a thing, so we make it a no op.

File: emsdk/patches/0001-JSDCE-Add-support-for-destructuring-declarations-204.patch
Comment (ryanking13) [inline emsdk/patches/0001-JSDCE-Add-support-for-destructuring-declarations-204.patch:1]: What is this patch for?

File: .circleci/config.yml [hunk 4]
```diff
@@ -22,7 +22,14 @@ orbs:
 
 jobs:
   build-core:
+    parameters:
+      disable_dylink:
+        description: The packages to be built.
```
Comment (ryanking13) [inline .circleci/config.yml:27]: description needs to be updated

**Review by ryanking13 — APPROVED** [review]
Thanks, @hoodmane!

File: emsdk/patches/0001-JSDCE-Add-support-for-destructuring-declarations-204.patch
Comment (hoodmane) [reply emsdk/patches/0001-JSDCE-Add-support-for-destructuring-declarations-204.patch:1]: https://github.com/emscripten-core/emscripten/issues/20393
